go_src/vagrant-vmware-utility/internal/service/port_forwarding.go

```diff
@@ -39,59 +40,94 @@ func (f *Forward) Activate() error {
 	}
 	f.Active = true
 
-	l, err := net.Listen(f.Fwd.Host.Type, f.Fwd.Host.String())
-	if err != nil {
-		f.logger.Error("failed to setup host listener", "host", f.Fwd.Host, "error", err)
-		return err
-	}
+	if f.Fwd.Host.Type == "" || strings.Contains(f.Fwd.Host.Type, "tcp") {
+		l, err := net.Listen("tcp", f.Fwd.Host.String())
+		if err != nil {
+			f.logger.Error("failed to setup host listener", "type", "tcp", "host", f.Fwd.Host, "error", err)
+			return err
+		}
 
-	go func() {
-		<-f.Ctx.Done()
-		l.Close()
-	}()
+		go func() {
+			<-f.Ctx.Done()
+			l.Close()
+		}()
 
-	f.logger.Debug("activated port forwarding", "fwd", f)
+		f.logger.Debug("activated port forward", "type", "tcp", "fwd", f)
 
-	go func() {
-		for {
-			conn, err := l.Accept()
-			if err != nil {
-				f.logger.Error("failed to accept incoming connection", "fwd", f, "error", err)
-				f.cancel()
-				return
-			}
+		go func() {
+			for {
+				conn, err := l.Accept()
+				if err != nil {
+					f.logger.Error("failed to accept incoming connection", "type", "tcp", "fwd", f, "error", err)
+					f.cancel()
+					return
+				}
 
-			target, err := net.Dial(f.Fwd.Guest.Type, f.Fwd.Guest.String())
-			if err != nil {
-				f.logger.Warn("failed to connect to guest", "guest", f.Fwd.Guest)
-				continue
+				target, err := net.Dial("tcp", f.Fwd.Guest.String())
+				if err != nil {
+					f.logger.Warn("failed to connect to guest", "type", "tcp", "guest", f.Fwd.Guest)
+					continue
+				}
+
+				ctx, completed := context.WithCancel(f.Ctx)
+				f.logger.Debug("initializing new connection stream", "type", "tcp", "fwd", f, "source", conn.RemoteAddr())
+				go f.stream(conn, target, completed, "tcp", "outgoing")
+				go f.stream(target, conn, completed, "tcp", "incoming")
+
+				go func() {
+					select {
+					case <-ctx.Done():
+					case <-f.Ctx.Done():
+					}
+					conn.Close()
+					target.Close()
+				}()
 			}
+		}()
+	}
 
-			ctx, completed := context.WithCancel(f.Ctx)
-			f.logger.Debug("initializing new connection stream", "fwd", f, "source", conn.RemoteAddr())
-			go f.stream(conn, target, completed)
-			go f.stream(target, conn, completed)
+	if f.Fwd.Host.Type == "" || strings.Contains(f.Fwd.Host.Type, "udp") {
```

Comment:
```suggestion
	if strings.Contains(f.Fwd.Host.Type, "udp") {
```

I think this shouldn't have the `||` bit. From the [Vagrant documentation](https://www.vagrantup.com/docs/networking/forwarded_ports#protocol)

> Either "udp" or "tcp". This specifies the protocol that will be allowed through the forwarded port. By default this is "tcp".

So, if udp is not specified as the type Vagrant should not forward using udp.

Reply:
:+1: originally when I wrote this it was with the intention of imitating vmware's default behavior (which is just forward everything) but the plugin internally will specify tcp by default when no value is provided and I didn't want to modify that behavior. Will back that bit out :slightly_smiling_face: